flang/test/Lower/OpenMP/lastprivate-allocatable.f90

```diff
@@ -36,3 +24,45 @@ program lastprivate_allocatable
 ! CHECK-NEXT:            }
 ! CHECK-NEXT:            omp.yield
 ! CHECK-NEXT:          }
+program lastprivate_allocatable
+  integer, allocatable :: a
+  integer :: i
+  ! a is unallocated here
+  !$omp parallel do lastprivate(a)
+  do i=1,1
+    a = 42
+  enddo
+  !$omp end parallel do
+  ! a should be allocated here
+end program
+
+! CHECK-LABEL:  func @_QPlastprivate_realloc()
+! CHECK:          %[[A:.*]]:2 = hlfir.declare %{{.*}} {fortran_attrs = #fir.var_attrs<allocatable>, uniq_name = "_QFlastprivate_reallocEa"} :
+! CHECK-SAME:       (!fir.ref<!fir.box<!fir.heap<!fir.array<?x!fir.complex<4>>>>>) ->
+! CHECK-SAME:       (!fir.ref<!fir.box<!fir.heap<!fir.array<?x!fir.complex<4>>>>>, !fir.ref<!fir.box<!fir.heap<!fir.array<?x!fir.complex<4>>>>>)
+! CHECK:          omp.parallel {
+! CHECK:            %[[A_PRIV:.*]]:2 = hlfir.declare %{{.*}} {fortran_attrs = #fir.var_attrs<allocatable>, uniq_name = "_QFlastprivate_reallocEa"} :
+! CHECK-SAME:         (!fir.ref<!fir.box<!fir.heap<!fir.array<?x!fir.complex<4>>>>>) ->
+! CHECK-SAME:         (!fir.ref<!fir.box<!fir.heap<!fir.array<?x!fir.complex<4>>>>>, !fir.ref<!fir.box<!fir.heap<!fir.array<?x!fir.complex<4>>>>>)
+! CHECK:            omp.sections {
+! CHECK:              omp.section {
+! CHECK:                fir.if %{{.*}} {
```

Comment:
Is this `if` correct? On my reading of the full generated code (omitted from the test case here) it seems to be testing whether the original (outside of the parallel region) copy of the variable is allocated. I would imagine that we should still perform the assignment even if it is not.

There is no such if statement in the `lastprivate_allocatable` testcase (the if statement is checking whether it is the final iteration of the loop). The "last" section is judged lexically so it doesn't need an if statement.

Reply:
This `if` was already being generated for allocatables in `copyVarHLFIR`, so I didn't want to change it.
It is indeed testing whether the original copy of the variable is allocated.
If I understood the standard correctly, it has this restriction:

> A variable that appears in a lastprivate clause must be deﬁnable.
(https://www.openmp.org/spec-html/5.2/openmpsu39.html)

The copy must also be allocated upon exit of the last section:

> If the original list item has the ALLOCATABLE attribute, the corresponding list item of which the value is assigned to the original item must have an allocation status of allocated upon exit from the sequentially last iteration or lexically last structured block sequence associated with a sections construct.

We don't check this, but since this is non-conforming code it should be OK. OTOH, following this logic, it would also be OK to not check if the original variable is allocated and let the behavior be undefined, at least for lastprivate.

> There is no such if statement in the lastprivate_allocatable testcase (the if statement is checking whether it is the final iteration of the loop).

Hmm, interesting, I hadn't noticed that. In fact, I think that `DataSharingProcessor::copyLastPrivateSymbol` could be almost entirely replaced by this new version of `copyHostAssociateVar/copyVarHLFIR`.

I don't know if, besides lastprivate, all other uses of `copyHostAssociateVar` can skip the `if allocated` part.
But if they do and If others agree we can remove it.

What do you think?

Reply:
> If the original list item does not have the POINTER attribute, its update occurs as if by intrinsic assignment unless it has a type bound procedure as a deﬁned assignment. 

If an allocated allocatable is assigned to an unallocated allocatable, the lhs is allocated as part of the assignment. So I think we should still do the assignment, even if the list item was unallocated.

Reply:
OK, I'll remove this `if` and adjust the tests.

BTW, according to the sentence that you have quoted above, when there is a `type bound procedure as a deﬁned assignment` it should be used, but it's not, because we set `AssignOp`'s `temporary_lhs` flag to false in `Bridge.cpp`.

There is a comment in the code saying that `temporary_lhs` also avoids finalization calls, but it seems to me we shouldn't avoid them:

> Finalization of a list item of a finalizable type or subobjects of a list item of a finalizable type occurs at the end of the region. The order in which any final subroutines for different variables of a finalizable type are called is unspecified.
(OMP 5.2, 5.3 List Item Privatization)

So I intend to also change `temporary_lhs` to false to fix these nonconformities. This is already being done in `DataSharingProcessor::copyLastPrivateSymbol`. Please let me know if you disagree.

Reply:
Good point!